File: pubnub-java/src/main/java/com/pubnub/api/models/consumer/pubsub/BasePubSubResult.java
package com.pubnub.api.models.consumer.pubsub;

import com.google.gson.JsonElement;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.ToString;

@Builder
@Getter
@ToString
@AllArgsConstructor
public class BasePubSubResult implements PNEvent {

    @Deprecated
    private String subscribedChannel;
    @Deprecated
    private String actualChannel;

    private String channel;
    private String subscription;

    private Long timetoken;

    private JsonElement userMetadata;

    private String publisher;

    protected BasePubSubResult(BasePubSubResult basePubSubResult) {
        this.subscribedChannel = basePubSubResult.subscribedChannel;
        this.actualChannel = basePubSubResult.actualChannel;
        this.channel = basePubSubResult.channel;
        this.subscription = basePubSubResult.subscription;
        this.timetoken = basePubSubResult.timetoken;
        this.userMetadata = basePubSubResult.userMetadata;
        this.publisher = basePubSubResult.publisher;
    }
}

File: core/src/main/kotlin/com/pubnub/api/models/consumer/pubsub/BasePubSubResult.kt
package com.pubnub.api.models.consumer.pubsub

import com.google.gson.JsonElement

interface PubSubResult : PNEvent {
    val channel: String
    val subscription: String?
    val timetoken: Long?
    val userMetadata: JsonElement?
    val publisher: String?
}

/**
 * @property channel The channel a PubNub API operation is related to.
 * @property subscription The subscription a PubNub API operation is related to.
 * @property timetoken Timetoken of the PubNub API operation in question.
 * @property userMetadata User metadata if any.
 * @property publisher The publisher of the PubNub API operation in question.
 */
data class BasePubSubResult(
    override val channel: String,
    override val subscription: String?,
    override val timetoken: Long?,
    override val userMetadata: JsonElement?,
    override val publisher: String?
) : PubSubResult
